### DISCUSSION TITLE
Can NextJS be used for (or handle) Partial injections?

### DISCUSSION THREAD

**Discussion by codinsonn** [vercel/next.js#19493] [Ideas]
## Can NextJS be used for (or handle) Partial injections? 

Other ways I could phrase this question:
- Has anyone succeeded in turning NextJS into a partial / component renderer?
- Could `pages/index.js` embed & hydrate mutiple html "partials" rendered from calls to 
`pages/render-component/[compName].js`?

## Example / What I Tried:

The  `_document.tsx` file only returns `<Main />` & `<NextScript />` for "partial" routes (omitting `<Html />`& `<Head />`)

<details>
<summary>pages/_document.tsx</summary>

```tsx
// @ts-nocheck
import React from 'react';
import Document, { Html, Head, Main, NextScript } from 'next/document';

/* --- <CustomDocument/> ---------------------------------------------------------------------- */

class CustomDocument extends Document {

    static async getInitialProps(ctx) {
      const initialProps = await Document.getInitialProps(ctx)
      return { ...initialProps }
    }
  
    render() {
        const { page } = this.props.__NEXT_DATA__; 
        
        // -- Render Component? --
       // -i- Only render component HTML & Scripts for partial injections

        if (page.includes('/render')) {
            return (
                <div>
                    <Main />
                    <NextScript />
                </div>
            );
        }

        // -- Render Page --

        return (
            <Html>
                <Head />
                <body>
                    <Main />
                    <NextScript />
                </body>
            </Html>
        );
    }
}

/* --- Exports ------------------------------------------------------------------------------ */

export default CustomDocument;

```

</details>

> The idea here is to enable injecting certain component into a page
> (e.g. via Twig macro's which call NextJS routes serverside, or other templating languages)

### What doesn't work out:

It seems that embedding multiple `<NextScript />`'s on a page, only result in one of those hydrating the entire page, deleting all other content.

<details>
<summary>pages/index.tsx</summary>

```tsx
import React from 'react';
// Components
import LabelTest from '../components/LabelTest';

/* --- <Home/> --------------------------------------------------------------------- */

const Home = (props) => {
    return (
        <div>
            <h1>
                <LabelTest label={props.body?.label} />
            </h1>
            <br />
            <div dangerouslySetInnerHTML={{ __html: props.component }} />
        </div>
    );
};

/* --- SSR Props ------------------------------------------------------------------------------ */

export const getServerSideProps = async (ctx) => {
    const component = 'SayHello';
    const componentProps = JSON.stringify({ greeting: 'World' });
    const res = await fetch(`http://localhost:3000/render-dynamic/${component}`, { method: 'POST', body: componentProps });
    const componentHTML = await res.text();
    console.log({ componentHTML }, ctx.req.host);
    return { props: { component: componentHTML } };
};

/* --- Exports ---------------------------------------------------------------------------------- */

export default Home;

```

</details>

Can this somehow be avoided?

### CodeSandbox & Minimal Repro Link

Codesandbox: https://codesandbox.io/s/nextjs-partials-e1jxi?file=/pages/index.js
Repo: https://github.com/codinsonn/nextjs-partials-codesandbox

**Comment by dpatel4-lat** [comment]
@codinsonn  any luck with this? We are looking for same.

**Reply by codinsonn** [reply]
I'm afraid not 😅 It works somewhat when you only render 1 partial on a page, but as soon as you try multiple, the others dissapear on hydration. Seems like NextJS expects there to only be one large component to hydrate.

What you CAN do, is create your own server side renderer for components / partials via an API route: e.g. using ReactDOM.renderToString() at '/pages/api/components/[component].js` and build a scraping script in your page load logic to hydrate each server rendered react html individually, but that would be a lot of custom work for something that doesn't really give you all the benefits of NextJS 😅
